### ISSUE TITLE
A safe PDF export

### ISSUE THREAD

**Issue by davidmehren** [hedgedoc/hedgedoc#965]
In 1.5, PDF export was disabled for security reasons. We still like to have that feature back, so this is the issue to discuss different ideas for a safe reimplementation.

**Current ideas:**
- Use pandoc
  - ➕ Probably safe
  - ➕ Supports many output formats in addition to PDF
  - ➖ Has native dependencies: Pandoc and a PDF-renderer (pdflatex, lualatex, xelatex, latexmk, tectonic, wkhtmltopdf, weasyprint, prince, context, or pdfroff)
  - ➖ Does not support any of our custom rendering features
- Use our renderer, but make it safe™
  - ➕ Supports all features
  - ❓ Safety unclear
    - PDF-renderer could be run as a separate service, e.g in a VM 
    
*If you have more ideas or aspects that are not yet in this list, please comment below!*

**Comment by DerMolly** [comment]
Not sure if this is really feasible as we would need to implement everything twice, but pandoc has filters to manipulate the AST and therefore the output. https://pandoc.org/filters.html

**Comment by belamu** [comment]
In the meantime, this comment shows how to export with Chrome/Chromium: https://github.com/hackmdio/codimd/issues/33#issuecomment-267508760

**Comment by InnayTool** [comment]
I think making the print pdf option in the browser more obvious would solve 90% of the users problems.
This can be done for example by switching to view mode and run `window.print()`.

**Comment by b3** [comment]
As I stated in #618 (https://github.com/hedgedoc/hedgedoc/issues/618#issuecomment-774382634) maybe you could check what has been done for https://github.com/osener/markup.rocks (demo page is at http://markup.rocks/). Said shortly they use a compilation of pandoc from Haskell to javascript thus having the benefit of the whole on pandoc.

**Comment by b3** [comment]
>       * heavy_minus_sign Does not support any of our custom rendering features

Where can I find a list of these specific custom rendering features?

What about implementing them as Pandoc filters (or [lua filters](https://pandoc.org/lua-filters.html)) as stated before?

>         * PDF-renderer could be run as a separate service, e.g in a VM

I think that it can also be done with Pandoc (some links are available on [Pandoc extra wiki page](https://github.com/jgm/pandoc/wiki/Pandoc-Extras#web-services-to-process-files-by-pandoc)).

**Comment by DerMolly** [comment]
> Where can I find a list of these specific custom rendering features?

Not sure we have a complete list somewhere, but most if not all should be showcased on ui-test.hedgedoc.org

**Comment by mrdrogdrog** [comment]
We're working on a list of features in GFM, CommonMark, HedgeDochMark 1 and HedgeDocMark2

**Comment by InnayTool** [comment]
@b3 I'm not sure if this setup is stable enough and it adds the Haskell dependency to the project.
In general that adds much complexity to the project.

**Comment by mrdrogdrog** [comment]
I'm against bundling pandoc into the new client. This would increase the packed size extremly for just one feature. And if the server generates the pdf with pandoc, then i don't see why we can't use the real pandoc instead.

**Comment by mrdrogdrog** [comment]
Regarding the print-to-pdf option: 

Yes it's possible and until we have a safe method to generate PDFs it's the only way to do that task.. but the big downside of the browser method is: You can't search through the texts because they are rendered as images.

**Comment by mrdrogdrog** [comment]
What about just using the browser pdf export?

**Comment by davidmehren** [comment]
I added the browser PDF export to the list.
If we add platform specific hints on how to use the print-feature (example: FF on Android does not seem to support printing, so we could show "use another browser") when the user presses the export button, this seems to be a promising solution.

**Comment by davidmehren** [comment]
So the question on the table is now: What use-cases does the browser PDF export *not* solve?

**Comment by SISheogorath** [comment]
First and foremost, as far as I can tell, the output is wrong on Firefox. For example for RevealJS:


https://cloud.shivering-isles.com/s/dgyzt63cEF8iAZp (Firefox)

https://cloud.shivering-isles.com/s/E8pJsAjxMKBcM5P (Chromium)

I can however say that Firefox became a lot better since the last time.

**Comment by mrdrogdrog** [comment]
It seems like Firefox isn't including the background graphics. Do you have to check a box for that like in chrome?

**Comment by ErikMichelson** [comment]
There is a checkbox at least:
![image](https://user-images.githubusercontent.com/52606896/109469976-10fe6c80-7a6f-11eb-9043-cda34c57c550.png)

**Comment by mrdrogdrog** [comment]
"use the print to pdf function of your browser!" Is safe, but not very intuitive. Imho it's bad UX to say "remember to check that box". That's the downside.

**Comment by SISheogorath** [comment]
@ErikMichelson I actually looked for that but couldn't find it in Firefox. Maybe I'm blind, but it doesn't speak for the browser function.

However, I still consider it a valid way to go, if the alternative is to put a browser in a cage on a server just to do the same thing.

**Comment by funkyfuture** [comment]
i also don't see why one software should solve all exotic and obsolete feature requests and increase complexity to achieve these.

if users want to convert markdown to pdf, let them use pandoc. possibly via a dedicated webservice that pulls data from HedgeDoc.

for users that are keen on creating printable contents there's LibeOffice and whatnot.

**Comment by SISheogorath** [comment]
PDF-export is not at all an exotic feature request. It definitely ranks in the top 5 of the most requested features. And while a lot of people have built workaround like using pandoc to process the markdown from HedgeDoc, a proper integrated version with good UX is preferred by many. So thinking about this feature and figuring it out properly in order to make people's life easier, is a good idea.

**Comment by b3** [comment]
The advantage of `pandoc` is the proven efficiency and the number of export possibilities as well as its extensibility. Enabling the use of a webservice offering a simple `pandoc` call will be nice (since such an export will be delegated, letting `hedgedoc` focus on more important work: synchronization and UX for instance).

For now, I am still looking at such a webservice though.

**Comment by funkyfuture** [comment]
nah, i'd file this great file format for printed media under the *obsolete* section in this context. HTML/Markdown are screen-oriented encodings (and while satisfying print-targettig css are rare, HedgeDoc doesn't support custom css). if a user wants to print a rendered Markdown document, the browser's print function should be sufficient. if you're promising a software that is suited to produce paged, printable documents, i'd personally consider that a hoax.

**Comment by DerMolly** [comment]
> HedgeDoc doesn't support custom css

Well technically it does.

> if you're promising a software that is suited to produce paged, printable documents, i'd personally consider that a hoax.

I don't think we intent to provide more than what was once part of this software, a (hopefully) simple PDF export…

**Comment by SISheogorath** [comment]
For those coming across this thread and wonder what was unsafe about the old PDF export: https://github.com/hedgedoc/hedgedoc/security/advisories/GHSA-pxxg-px9v-6qf3

**Comment by almereyda** [comment]
Here we can possibly also take inspiration from

* [Vivliostyle — Enjoy CSS Typesetting!](https://vivliostyle.org/)

via

* [Shinyu Murakami on Vivliostyle and the state of webbased publishing | Fidus Writer](https://www.fiduswriter.org/2018/11/09/shinyu-murakami-on-vivliostyle-and-the-state-of-webbased-publishing/)

which each, Fidus Writer and Vivliostyle, approach similar objectives as HedgeDoc here.

Fidus Writer then, in the end, also only opens the browser's print dialogue and provides an adapted CSS with a media query. In direction of

> For now, I am still looking at such a webservice though.

there was once

* [mrded/pandoc-as-a-service: If you need to convert text from one markup format into another, pandoc-as-a-service is your swiss-army knife.](https://github.com/mrded/pandoc-as-a-service)

which seems to be a suitable PoC for porting into a more long-term oriented architecture.

**Comment by davidmehren** [comment]
Thanks for your ideas, @almereyda!

I had a look at vivliostyle, it boils down to the CLI [running Chromium with pupeteer](https://github.com/vivliostyle/vivliostyle-cli/blob/main/src/pdf.ts#L57-L66). We can't do that, at least not on the server, as it has the same security-problems as the PDF-generation we had to ditch with HedgeDoc 1.5. It *might* be an option for a future HedgeDoc 2 CLI though.

wkhtmltopdf is also more or less a browser, so the same security considerations apply.

**Comment by schyzophrene-asynchrone** [comment]
Using `pandoc` with any flavour of latex as backend is a bad idea, as this would enable any latex command to be executed, giving any user access to all files readable by the server’s user which ran `pandoc` (through the `\input` command).

**Comment by almereyda** [comment]
There's a German article at https://www.heise.de/hintergrund/Auf-Knopfdruck-PDFs-in-Anwendungen-erstellen-mit-HTML-und-CSS-Modul-6457863.html?seite=all which might provide hints at how others solve this quest, but basically it says to use Print CSS.

**Comment by tarleb** [comment]
A good solution could be [Typst](https://typst.app), as it's small, fast, and supported by pandoc. Typst can even be compiled to wasm and thus be run in the browser, but that would require custom code to convert to Typst's markup. (Pandoc can run in the browser, too, but it's not officially supported, and the binaries are rather large.)
